### DISCUSSION TITLE
server does not allow insecure connections, client must use SSL/TLS

### DISCUSSION THREAD

**Discussion by NLanese** [planetscale/discussion#354] [Q&A] [answered]
Following all of the PlanetScale / Prism installation instructions gets me close, but whenever I try to run `npx prisma db push` I get the following error: `server does not allow insecure connections, client must use SSL/TLS` 
Every solution I've seen is vastly different, some even directly refuting each other. I've tried adding `?tls=true` to the DATABASE_URL, and that does not work. Adding more complicated additions  to the DATABASE_URL just gives me a different error, just telling me the database cannot be found. I have no idea how to set up TLS, yet many docs make it seem like I should already be set to use it. I sincerely have no idea what is going on, and I cannot even get started using PlanetScale since I can't even set up my db

**Comment by fooforge** [comment, marked as answer]
@NLanese No worries, SSL/TLS can be confusing.

The first thing you need to check is if your client has the common CA (certificate authority) certificates installed, sometimes also called the CA root store. These certificates are usually all in a single file in `.pem` format. They are necessary in order for your computer to be able to verify our certificate which, in turn, allows it to establish a secure connection with PlanetScale.

On OS X, these certificates are in `/etc/ssl/cert.pem`. On Linux, it depends on the distribution ([see our documentation](https://planetscale.com/docs/concepts/secure-connections#ca-root-configuration) for some common paths). Windows is a bit of a special snowflake. It has an internal database where these certificates are stored, but some database drivers cannot communicate with Windows' CryptoAPI (the interface with which you can access Windows' internal database) and require you to point the driver to a file instead.
In that case, you can download a bundled version of these CA certificates curated by [Mozilla's CA Certificate Program](https://wiki.mozilla.org/CA) from https://curl.se/docs/caextract.html, add the file to your project, and point your database driver to it.

Generally, for the secure connection to succeed, all you really need to do is to add `sslaccept=strict` to your connection string. This parameter ensures that you can only open a secure connection to your database and that the connection fails if the certificate could not be verified. It protects you from establishing an unsecured connection which would make you susceptible to man-in-the-middle attacks.

When on Windows and when you have downloaded the CA bundle, the only other parameter you may need to add is `sslcert=/path/to/the/ca-certificates.pem` to point your database driver to the downloaded file as described earlier.

On OS X or Linux, your `schema.prisma` should look like this:

```js
datasource db {
  provider = "mysql"
  url      = "mysql://[USERNAME]:[PASSWORD]@[HOSTNAME]:3306/[DATABASE]?sslaccept=strict"
}
```

And, on Windows like this:

```js
datasource db {
  provider = "mysql"
  url      = "mysql://[USERNAME]:[PASSWORD]@[HOSTNAME]:3306/[DATABASE]?sslaccept=strict&sslcert=path/to/cacert.pem"
}
```

All other parameters such as `sslidentity` or `sslpassword` which are lined out in [Prisma's documentation](https://www.prisma.io/docs/concepts/database-connectors/mysql#configuring-an-ssl-connection) are not important in the context of PlanetScale and you will not need them. Same goes for the `openssl` command in that document.

Ultimately, I think, this is where a lot of the confusion stems from. SSL/TLS, or really OpenSSL, also allows you to act as your own certificate authority (CA), meaning you can issue your own client certificates instead of using one of the common CAs such as Let's Encrypt, Comodo or Digicert. When you issue your own certificates, however, those aren't automatically trusted by the browser or the OS. You use these for scenarios where you need two internal services to communicate through an encrypted channel, for example.

All that is irrelevant in the context of PlanetScale, however. To be able to establish a secure connection with us, all your computer needs to be able to do is to verify the legitimacy of our SSL certificate for `*.planetscale.com` and for that it needs to be able to consult your computer's local certificate authority (CA) store. There, it will find Let's Encrypt's root certificates which your computer then uses to validate our certificate.

I hope this helps! If not, please share some more information about your client, the exact error message that you're seeing and share your connection string with me with any secrets obscured.

**Reply by NLanese** [reply]
Absolutely amazing! Thank you!!! Weird that the one PlanetScale gave me itself didn't work. Thank you again!
